src/qminer/qminer_storage.cpp

```diff
@@ -4193,333 +4262,520 @@ PJsonVal TStorePbBlob::GetFieldJsonVal(const uint64& RecId, const int& FieldId)
+    }
+        
+    // set new value
     if (FieldLocV[FieldId] == TStoreLoc::slDisk) {
-        TPgBlobPt& PgPt = RecIdBlobPtH.GetDat(RecId);
-        TThinMIn min = DataBlob->Get(PgPt);
         SerializatorCache->SetFieldTm(min.GetBfAddrChar(), min.Len(), FieldId, Tm);
-        DataBlob->SetDirty(PgPt);
     } else {
-        TPgBlobPt& PgPt = RecIdBlobPtHMem.GetDat(RecId);
-        TThinMIn min = DataMem->Get(PgPt);
         SerializatorMem->SetFieldTm(min.GetBfAddrChar(), min.Len(), FieldId, Tm);
-        DataMem->SetDirty(PgPt);
     }
+
+    // index the new value in the updated memory buffer
+    RecIndexer.IndexRec(min.GetMemBase(), RecId, FieldId, *FieldSerializator);
 }
 /// Set field value using field id (default implementation throws exception)
 void TStorePbBlob::SetFieldTmMSecs(const uint64& RecId, const int& FieldId, const uint64& TmMSecs) {
+    // special case if field is primary field
+    if (FieldId == PrimaryFieldId) {
+        // it is, make sure new value does not exist yet
+        if (PrimaryTmMSecsIdH.IsKey(TmMSecs) && PrimaryTmMSecsIdH.GetDat(TmMSecs) != RecId) {
+            throw TQmExcept::New("[TStorePbBlob::SetFieldTmMSecs] Primary key '" + TUInt64::GetStr(TmMSecs) +
+                "' being set to field '" + GetFieldNm(FieldId) + "' already taken.");
+        }
+    }
+    // get the memory containig the field for the record
+    TThinMIn min = GetEditableField(RecId, FieldId);
+
+    // if we are indexing the field and the value is nonnull, first deindex the old value
+    TRecSerializator* FieldSerializator = GetFieldSerializator(FieldId);
+    if (RecIndexer.HasIndexKey(FieldId) && !IsFieldNull(RecId, FieldId)) {
+        RecIndexer.DeindexRec(min.GetMemBase(), RecId, FieldId, *FieldSerializator);
+    }
+    if (FieldId == PrimaryFieldId) { DelPrimaryFieldMSecs(RecId, FieldSerializator->GetFieldTmMSecs(min, FieldId)); }
+
+    // set new value
     if (FieldLocV[FieldId] == TStoreLoc::slDisk) {
-        TPgBlobPt& PgPt = RecIdBlobPtH.GetDat(RecId);
-        TThinMIn min = DataBlob->Get(PgPt);
         SerializatorCache->SetFieldTmMSecs(min.GetBfAddrChar(), min.Len(), FieldId, TmMSecs);
-        DataBlob->SetDirty(PgPt);
     } else {
-        TPgBlobPt& PgPt = RecIdBlobPtHMem.GetDat(RecId);
-        TThinMIn min = DataMem->Get(PgPt);
         SerializatorMem->SetFieldTmMSecs(min.GetBfAddrChar(), min.Len(), FieldId, TmMSecs);
-        DataMem->SetDirty(PgPt);
     }
+
+    // index the new value in the updated memory buffer
+    RecIndexer.IndexRec(min.GetMemBase(), RecId, FieldId, *FieldSerializator);
+    if (FieldId == PrimaryFieldId) { SetPrimaryFieldMSecs(RecId, TmMSecs); }
 }
 /// Set field value using field id (default implementation throws exception)
 void TStorePbBlob::SetFieldNumSpV(const uint64& RecId, const int& FieldId, const TIntFltKdV& SpV) {
+    TRecSerializator* FieldSerializator = GetFieldSerializator(FieldId);
```

Comment:
This code repeats a lot, it would be good to abstract it a bit and make it more compact. Not necessary for this commit.